plugin/hpsa/hpsa.py

```diff
@@ -573,18 +573,28 @@ def _hp_ld_to_lsm_vol(hp_ld, pool_id, sys_id, ctrl_num, array_num,
         """
         ld_num = hp_ld_name[len("Logical Drive: "):]
         vpd83 = hp_ld['Unique Identifier'].lower()
-        # No document or command output indicate block size
-        # of volume. So we try to read from linux kernel, if failed
-        # try 512 and roughly calculate the sector count.
-        device = Device.from_device_file(_CONTEXT, hp_ld['Disk Name'])
-        vol_name = "%s: /dev/%s" % (hp_ld_name, device.sys_name)
-        attributes = device.attributes
-        try:
-            block_size = attributes.asint("queue/logical_block_size")
-            num_of_blocks = attributes.asint("size")
-        except (KeyError, UnicodeDecodeError, ValueError):
-            block_size = 512
-            num_of_blocks = int(_hp_size_to_lsm(hp_ld['Size']) / block_size)
+        hp_ld_path = ''
+        if 'Disk Name' in hp_ld:
+            hp_ld_path = hp_ld['Disk Name']
+        else:
+            hp_ld_path_list = LocalDisk.vpd83_search(vpd83)
+            if len(hp_ld_path_list) > 0:
+                hp_ld_path = hp_ld_path_list[0]
+
+        if hp_ld_path != '':
+            device = Device.from_device_file(_CONTEXT, hp_ld_path)
+            vol_name = "%s: /dev/%s" % (hp_ld_name, device.sys_name)
+            attributes = device.attributes
+            # No document or command output indicate block size
+            # of volume. So we try to read from linux kernel, if failed
+            # try 512 and roughly calculate the sector count.
+            try:
+                block_size = attributes.asint("queue/logical_block_size")
+                num_of_blocks = attributes.asint("size")
+            except (KeyError, UnicodeDecodeError, ValueError):
+                block_size = 512
+                num_of_blocks = int(
+                    _hp_size_to_lsm(hp_ld['Size']) / block_size)
```

Comment:
Please try this patch: https://github.com/cathay4t/libstoragemgmt/commit/4a179265d748c03288c68d2b309095e4cd929cf9 
```patch
From 4a179265d748c03288c68d2b309095e4cd929cf9 Mon Sep 17 00:00:00 2001
From: Gris Ge <fge@redhat.com>
Date: Wed, 20 Jul 2016 09:33:41 +0800
Subject: [PATCH] HPSA Plugin: Use LocalDisk API to retrieve volume path

Fixing issue https://github.com/libstorage/libstoragemgmt/issues/167

Signed-off-by: Gris Ge <fge@redhat.com>
---
 plugin/hpsa/hpsa.py | 25 ++++++++++++++++---------
 1 file changed, 16 insertions(+), 9 deletions(-)

diff --git a/plugin/hpsa/hpsa.py b/plugin/hpsa/hpsa.py
index 648e1fc..3cf93ce 100644
--- a/plugin/hpsa/hpsa.py
+++ b/plugin/hpsa/hpsa.py
@@ -576,15 +576,22 @@ def _hp_ld_to_lsm_vol(hp_ld, pool_id, sys_id, ctrl_num, array_num,
         # No document or command output indicate block size
         # of volume. So we try to read from linux kernel, if failed
         # try 512 and roughly calculate the sector count.
-        device = Device.from_device_file(_CONTEXT, hp_ld['Disk Name'])
-        vol_name = "%s: /dev/%s" % (hp_ld_name, device.sys_name)
-        attributes = device.attributes
-        try:
-            block_size = attributes.asint("queue/logical_block_size")
-            num_of_blocks = attributes.asint("size")
-        except (KeyError, UnicodeDecodeError, ValueError):
-            block_size = 512
-            num_of_blocks = int(_hp_size_to_lsm(hp_ld['Size']) / block_size)
+        block_size = 512
+        num_of_blocks = int(_hp_size_to_lsm(hp_ld['Size']) / block_size)
+        vol_name = hp_ld_name
+
+        if len(vpd83) > 0:
+            blk_paths = LocalDisk.vpd83_search(vpd83)
+            if len(blk_paths) > 0:
+                blk_path = blk_paths[0]
+                vol_name += ": %s" % " ".join(blk_paths)
+                device = Device.from_device_file(_CONTEXT, blk_path)
+                attributes = device.attributes
+                try:
+                    block_size = attributes.asint("queue/logical_block_size")
+                    num_of_blocks = attributes.asint("size")
+                except (KeyError, UnicodeDecodeError, ValueError):
+                    pass
 
         if 'Failed' in hp_ld['Status']:
             admin_status = Volume.ADMIN_STATE_DISABLED
-- 
2.9.0

```